src/bridge/cockpitcgroupsamples.c

```diff
@@ -21,119 +21,207 @@
 
 #include "cockpitcgroupsamples.h"
 
+#include <errno.h>
+#include <fcntl.h>
 #include <fts.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
 
-const gchar *cockpit_cgroup_memory_root = "/sys/fs/cgroup/memory";
-const gchar *cockpit_cgroup_cpuacct_root = "/sys/fs/cgroup/cpuacct";
+const char *cockpit_cgroupv1_memory_root = "/sys/fs/cgroup/memory";
+const char *cockpit_cgroupv1_cpuacct_root = "/sys/fs/cgroup/cpuacct";
+const char *cockpit_cgroupv2_root = "/sys/fs/cgroup";
 
-static double
-read_double (const gchar *prefix,
-             const gchar *suffix)
+static const char*
+read_file (int dirfd,
+           const char *cgroup,
+           const char *fname)
 {
-  gchar *path = NULL;
-  gchar *file_contents = NULL;
-  GError *error = NULL;
-  double ret;
-  gsize len;
-
-  path = g_build_filename (prefix, suffix, NULL);
-  if (!g_file_get_contents (path, &file_contents, &len, &error))
+  static char buf[256];
+  int fd = -1;
+  ssize_t len;
+  const char *ret = NULL;
+
+  fd = openat (dirfd, fname, O_RDONLY);
+  if (fd < 0)
     {
-      if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT) ||
-          g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NODEV))
-        g_debug ("samples file not found: %s", path);
+      if (errno == ENOENT || errno == ENODEV)
+        g_debug ("samples file not found: %s/%s", cgroup, fname);
       else
-        g_message ("error loading file: %s: %s", path, error->message);
-      g_error_free (error);
-      ret = -1;
+        g_message ("error opening file: %s/%s: %m", cgroup, fname);
+      return NULL;
     }
-  else
+
+  /* don't do fancy retry/error handling here -- we know what cgroupfs attributes look like,
+   * it's a virtual file system (does not block/no multiple reads), and it's ok to miss
+   * one sample due to EINTR or some race condition */
+  len = read (fd, buf, sizeof buf);
+  if (len < 0)
     {
-      ret = g_ascii_strtod (file_contents, NULL);
+      g_message ("error loading file: %s/%s: %m", cgroup, fname);
+      goto out;
     }
+  if (len >= sizeof buf)
+    {
+      g_warning ("cgroupfs value %s/%s is too large", cgroup, fname);
+      goto out;
+    }
+  buf[len] = '\0';
+  ret = buf;
 
-  g_free (file_contents);
-  g_free (path);
-
+out:
+  if (fd >= 0)
+    close (fd);
   return ret;
 }
 
+
+static gint64
+read_int64 (int dirfd,
+            const char *cgroup,
+            const char *fname)
+{
+  const char *contents = read_file (dirfd, cgroup, fname);
+  if (contents == NULL)
+      return -1;
+  return atoll (contents);
```

Comment:
We're talking to the kernel, of course, so I'm not sure how much error checking you want, ...

but maybe the file format changes or something?  We have warnings above for missing files...

Reply:
I added error checking for read_prefixed_int64(), but reverted it for general read_int64(). There are a lot of attributes whose value is "max" for example, and I don't want to hardcode a list of expected non-numbers, but just treat them as absent. So I went back to atoll() and ignoring a zero result where that is invalid.